File: clients/python/src/slate/_details.py
"""The stable reason token out of a status's `grpc-status-details-bin`.

The head node puts a `google.rpc.ErrorInfo` in every status's details —
`crates/slate-server/src/status.rs` — because the status code alone is
many-to-one: `UNAVAILABLE` covers a fenced writer, a stale replica, no replica
and a failed object-store call, and `RESOURCE_EXHAUSTED` covers a join build,
a group count, a distinct count, a sort and a predicate write that matched more
rows than it may return. The token is how a caller tells them apart without
matching on prose.

# Why this builds its own descriptors instead of importing `google.rpc`

The obvious implementation is to generate `google/rpc/status.proto` and
`error_details.proto` into this package the way `slate/v1/records.proto` is
generated, and import the result. **That breaks `import slate` outright** for
anyone who also has `googleapis-common-protos` installed — which is most
people, since `grpcio-status`, every `google-cloud-*` package and the OTLP
exporter all depend on it.

Generated modules register themselves in protobuf's *default* descriptor pool
under the proto's own path, and two files with that path whose descriptors are
not identical are a hard error at import time:

    TypeError: Couldn't build proto file into descriptor pool:
    duplicate file name google/rpc/error_details.proto

Identical copies are tolerated, which is what makes this trap quiet: this
repository's `error_details.proto` declares `ErrorInfo` alone, upstream's
declares ten messages, so the descriptors differ and the collision is certain
rather than possible. It was demonstrated before this module was written.

So the three messages are declared here, in a pool of this module's own, under
a file name nothing else can claim. Protobuf is structural — field numbers and
wire types are the contract, names are not — so these decode the head node's
bytes exactly as generated `google.rpc` classes would. The cost is that the
declarations below must stay in step with `proto/google/rpc/` by hand; they are
seven fields that have not changed since 2015, and `test_details.py` decodes a
blob built from the real thing.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Final

from google.protobuf import descriptor_pb2, descriptor_pool, message_factory

#: The `Any.type_url` the head node packs its `ErrorInfo` under.
#: `status.rs::ERROR_INFO_URL`.
ERROR_INFO_URL: Final = "type.googleapis.com/google.rpc.ErrorInfo"

_Field = descriptor_pb2.FieldDescriptorProto

_pool = descriptor_pool.DescriptorPool()
_file = descriptor_pb2.FileDescriptorProto()
# Neither the file name nor the package may be one anything else could declare,
# which is the whole point of the module docstring above.
_file.name = "slate/_status_details.proto"
_file.package = "slate._rpc"
_file.syntax = "proto3"

# `google.protobuf.Any`: type_url = 1, value = 2.
_any = _file.message_type.add(name="Any")
_any.field.add(
    name="type_url", number=1, type=_Field.TYPE_STRING, label=_Field.LABEL_OPTIONAL
)
_any.field.add(
    name="value", number=2, type=_Field.TYPE_BYTES, label=_Field.LABEL_OPTIONAL
)

# `google.rpc.Status`: code = 1, message = 2, details = 3 (repeated Any).
_status = _file.message_type.add(name="Status")
_status.field.add(
    name="code", number=1, type=_Field.TYPE_INT32, label=_Field.LABEL_OPTIONAL
)
_status.field.add(
    name="message", number=2, type=_Field.TYPE_STRING, label=_Field.LABEL_OPTIONAL
)
_status.field.add(
    name="details",
    number=3,
    type=_Field.TYPE_MESSAGE,
    label=_Field.LABEL_REPEATED,
    type_name=".slate._rpc.Any",
)

# `google.rpc.ErrorInfo`: reason = 1, domain = 2, metadata = 3.
_info = _file.message_type.add(name="ErrorInfo")
_info.field.add(
    name="reason", number=1, type=_Field.TYPE_STRING, label=_Field.LABEL_OPTIONAL
)
_info.field.add(
    name="domain", number=2, type=_Field.TYPE_STRING, label=_Field.LABEL_OPTIONAL
)
# The metadata map, which used to be left undeclared on the grounds that
# surfacing it "means deciding what a client promises about keys that vary per
# variant". That was right when every variant invented its own keys. It is no
# longer the whole story: the check-violation keys are specified — `violations`
# is a count and `check.N`, `column.N`, `message.N` are indexed from zero — so
# there is exactly one shape a client can promise something about.
#
# So the map is parsed and *not* exposed raw. `check_failures_of` reads the one
# documented shape into typed values and nothing hands a caller the dictionary,
# which keeps the original objection answered rather than overruled: a key this
# client has no contract for still reaches nobody.
#
# A `map<string, string>` on the wire is a repeated message of key/value pairs
# with the `map_entry` option set, which is what this builds by hand.
_entry = _info.nested_type.add(name="MetadataEntry")
_entry.field.add(
    name="key", number=1, type=_Field.TYPE_STRING, label=_Field.LABEL_OPTIONAL
)
_entry.field.add(
    name="value", number=2, type=_Field.TYPE_STRING, label=_Field.LABEL_OPTIONAL
)
_entry.options.map_entry = True
_info.field.add(
    name="metadata",
    number=3,
    type=_Field.TYPE_MESSAGE,
    label=_Field.LABEL_REPEATED,
    type_name=".slate._rpc.ErrorInfo.MetadataEntry",
)

_pool.Add(_file)
_Status = message_factory.GetMessageClass(_pool.FindMessageTypeByName("slate._rpc.Status"))
_ErrorInfo = message_factory.GetMessageClass(
    _pool.FindMessageTypeByName("slate._rpc.ErrorInfo")
)


def reason_of(blob: bytes) -> str:
    """The `ErrorInfo.reason` in a `grpc-status-details-bin` blob, or `""`.

    `""` for a blob that carries no `ErrorInfo`, and for one that does not
    parse. A client that raised while building an error message would replace
    the server's failure with its own, which is strictly worse than losing a
    token: the caller would no longer know why the call failed at all.

    The `ErrorInfo.metadata` map is not returned *here*. The token alone is
    what `errors.py` needs to branch below a status code. See
    `check_failures_of` for the one part of that map this client reads.
    """
    # `Any`, because these classes are built at import from a descriptor and
    # `message_factory` types them as the base `Message`: the field names below
    # exist at runtime and are invisible to a type checker, which is the price
    # of not importing `google.rpc`. The field names are pinned by the tests
    # rather than by the type checker, which is the weaker of the two guarantees and is
    # said here rather than left for a reader to discover.
    try:
        status: Any = _Status()
        status.ParseFromString(blob)
        for detail in status.details:
            if detail.type_url == ERROR_INFO_URL:
                info: Any = _ErrorInfo()
                info.ParseFromString(detail.value)
                return str(info.reason)
    except Exception:  # pragma: no cover - see the docstring
        return ""
    return ""


@dataclass(frozen=True)
class CheckFailure:
    """One `CHECK` a refused row violated.

    The point of the whole thing: a form can put `message` next to `column`
    without parsing it out of the status text, which is the contract a caller
    would otherwise have to invent — and which would last until somebody
    reworded a sentence.
    """

    #: The constraint's name, as the schema declares it.
    check: str
    #: The column it is about, or `None` for one spanning several. A caller
    #: with nowhere to put it shows it beside the form rather than a field.
    column: str | None
    #: The sentence to show, or `None` where the schema wrote none.
    message: str | None


def check_failures_of(blob: bytes) -> list[CheckFailure]:
    """Every check a refused write violated, in declaration order.

    Empty for any failure that is not a check violation, and for a blob that
    does not parse — a client that raised while building an error message
    would replace the server's failure with its own, which is the reasoning
    `reason_of` already gives.

    Reads the indexed keys (`check.0`, `column.0`, `message.0`, …) rather than
    the unindexed pair, because the unindexed one is only the *first* failure
    and this is the call that exists to return all of them. A server old
    enough to send only the unindexed pair yields one failure here, which is
    the honest reading of what it said.
    """
    try:
        status: Any = _Status()
        status.ParseFromString(blob)
        for detail in status.details:
            if detail.type_url != ERROR_INFO_URL:
                continue
            info: Any = _ErrorInfo()
            info.ParseFromString(detail.value)
            if str(info.reason) != "CHECK_VIOLATION":
                return []
            data = dict(info.metadata)
            # The count, not the key set: a message or column may legitimately
            # be absent, so counting `check.N` keys would be right and
            # counting the map's size would not. The server sends `violations`
            # for exactly this.
            try:
                total = int(data.get("violations", ""))
            except ValueError:
                # Old enough to send the unindexed pair and no count. One
                # failure is what that means.
                name = data.get("check")
                if name is None:
                    return []
                return [
                    CheckFailure(name, data.get("column"), data.get("message"))
                ]
            out = []
            for at in range(total):
                name = data.get(f"check.{at}")
                if name is None:
                    # A gap means the server and this reader disagree about
                    # the shape. Returning the prefix would be a quiet lie
                    # about how many rules the row broke.
                    return []
                out.append(
                    CheckFailure(name, data.get(f"column.{at}"), data.get(f"message.{at}"))
                )
            return out
    except Exception:  # pragma: no cover - see the docstring
        return []
    return []

File: clients/typescript/src/details.ts
/**
 * The stable reason token out of a status's `grpc-status-details-bin`.
 *
 * The head node puts a `google.rpc.ErrorInfo` in every status's details —
 * `crates/slate-server/src/status.rs` — because the status code alone is
 * many-to-one: `UNAVAILABLE` covers a fenced writer, a stale replica, no
 * replica and a failed object-store call, and `RESOURCE_EXHAUSTED` covers a
 * join build, a group count, a distinct count, a sort, and a predicate write
 * that matched more rows than it may hand back. The token is how a caller
 * tells them apart without matching on prose.
 *
 * The two `.proto` files this reads have been vendored in `proto/google/rpc/`
 * since the server grew rich errors, and until now nothing loaded them. They
 * are already listed in `files`, so they ship with the package and this adds
 * nothing to what is published.
 */

import * as path from "node:path";
// A default import, not `import * as`. protobufjs is CommonJS, and under
// NodeNext a namespace import of a CommonJS module gives the module namespace
// rather than `module.exports` — so `protobuf.Root` is `undefined` and the
// first call fails with "Root is not a constructor". Caught by the conformance
// runner rather than by the build: the namespace shape type-checks, and the
// decoder swallows its own failures by design, so the only symptom was this
// client reporting an empty token while the other two reported a real one.
import protobuf from "protobufjs";

import { directoryOf, findUpContaining } from "./paths.js";

/** The `Any.type_url` the head node packs its `ErrorInfo` under. */
export const ERROR_INFO_URL = "type.googleapis.com/google.rpc.ErrorInfo";

/** Where gRPC puts a status's `google.rpc.Status`, per the gRPC spec. */
export const DETAILS_KEY = "grpc-status-details-bin";

type StatusMessage = {
  details?: ReadonlyArray<{ type_url?: string; value?: Uint8Array }>;
};

let cached: { status: protobuf.Type; info: protobuf.Type } | undefined;
/**
 * Loaded once, on the first failure rather than at import.
 *
 * A client that read two `.proto` files off disk to import is a client that
 * pays for rich errors on a run that never has one, and `client.ts` loads its
 * own descriptor the same way and for the same reason.
 */
function types(): { status: protobuf.Type; info: protobuf.Type } | undefined {
  if (!cached) {
    try {
      const base = protoRoot();
      const root = new protobuf.Root();
      // `status.proto` imports `google/protobuf/any.proto`, which protobufjs
      // resolves from its own bundled copy; every other import resolves under
      // the vendored tree.
      root.resolvePath = (origin, target) =>
        target.startsWith("google/protobuf/")
          ? protobuf.Root.prototype.resolvePath(origin, target)
          : path.join(base, target);
      root.loadSync(["google/rpc/status.proto", "google/rpc/error_details.proto"], {
        keepCase: true,
      });
      cached = {
        status: root.lookupType("google.rpc.Status"),
        info: root.lookupType("google.rpc.ErrorInfo"),
      };
    } catch {
      // Losing the token is a degradation; throwing here would replace the
      // server's failure with this client's, which is not.
      return undefined;
    }
  }
  return cached;
}

let root: string | undefined;
/**
 * Where the vendored protos are.
 *
 * Found by looking for the file rather than counting `..` segments, for the
 * reason `client.ts` gives about its own `PROTO_ROOT`: this module runs from
 * `src/` in the repository, from `dist/` once built and from a package root
 * once installed, and a fixed depth is right in exactly one of the three. The
 * helpers live in `paths.ts` rather than in `client.ts` so that reusing them
 * here does not make `client.ts` and `errors.ts` import each other.
 */
function protoRoot(): string {
  if (root === undefined) {
    root = path.join(
      findUpContaining(
        directoryOf(import.meta.url),
        [path.join("proto", "google", "rpc", "error_details.proto")],
        "bundled proto directory",
      ),
      "proto",
    );
  }
  return root;
}

/**
 * The `ErrorInfo.reason` in a `grpc-status-details-bin` blob, or `""`.
 *
 * `""` for a blob carrying no `ErrorInfo` and for one that does not parse.
 *
 * `ErrorInfo.metadata` is not returned *here*. The token alone is what lets
 * `errors.ts` branch below a status code, and it is what three clients can
 * agree on. See {@link checkFailuresOf} for the one part of that map this
 * client reads.
 */
export function reasonOf(blob: Uint8Array): string {
  const loaded = types();
  if (!loaded) return "";
  try {
    const decoded = loaded.status.decode(blob) as unknown as StatusMessage;
    for (const detail of decoded.details ?? []) {
      if (detail.type_url !== ERROR_INFO_URL || !detail.value) continue;
      const info = loaded.info.decode(detail.value) as unknown as { reason?: string };
      return info.reason ?? "";
    }
  } catch {
    return "";
  }
  return "";
}

/** One `CHECK` a refused row violated. */
export interface CheckFailure {
  /** The constraint's name, as the schema declares it. */
  readonly check: string;
  /**
   * The column it is about, or `""` for a check spanning several.
   *
   * `""` rather than `undefined`, so that a caller rendering it into a form
   * writes `failure.column` and not a null check: a check over two columns has
   * no single field to hang the message on, and naming either would put the
   * sentence beside the wrong input.
   */
  readonly column: string;
  /** The sentence to show, or `""` where the schema wrote none. */
  readonly message: string;
}

type ErrorInfoMessage = { reason?: string; metadata?: Record<string, string> };

/**
 * Every `CHECK` a refused write violated, in declaration order.
 *
 * Empty for any failure that is not a check violation — which is almost all of
 * them — and for a blob that does not parse, for the reason {@link reasonOf}
 * gives about never throwing out of an error path.
 *
 * This reads `ErrorInfo.metadata`, which `reasonOf`'s comment used to say was
 * deliberately never exposed. That objection was about keys that vary per
 * variant, and it still holds: the check-violation keys are the one *specified*
 * shape — `violations` is a count and `check.N`, `column.N`, `message.N` are
 * indexed from zero — so this reads them into typed values and hands nobody the
 * raw dictionary. A key this client has no contract for still reaches no
 * caller.
 *
 * Counts up from `violations` rather than walking the map for `check.*`,
 * because the metadata is string-keyed: `check.10` sorts between `check.1` and
 * `check.2`, so a map walk is right for nine failures and wrong for eleven.
 *
 * Returns nothing rather than a prefix when the count and the keys disagree. A
 * caller shown two failures for a row that broke three fixes two fields,
 * resubmits and is refused again — the round-trip-per-field behaviour this
 * exists to remove.
 */
export function checkFailuresOf(blob: Uint8Array): CheckFailure[] {
  const loaded = types();
  if (!loaded) return [];
  try {
    const decoded = loaded.status.decode(blob) as unknown as StatusMessage;
    for (const detail of decoded.details ?? []) {
      if (detail.type_url !== ERROR_INFO_URL || !detail.value) continue;
      const info = loaded.info.decode(detail.value) as unknown as ErrorInfoMessage;
      if (info.reason !== "CHECK_VIOLATION") return [];
      const data = info.metadata ?? {};
      const total = Number.parseInt(data["violations"] ?? "", 10);
      if (!Number.isInteger(total)) {
        // A server old enough to send the unindexed pair and no count. One
        // failure is the honest reading of what it said.
        const name = data["check"];
        if (!name) return [];
        return [
          { check: name, column: data["column"] ?? "", message: data["message"] ?? "" },
        ];
      }
      const out: CheckFailure[] = [];
      for (let at = 0; at < total; at += 1) {
        const name = data[`check.${at}`];
        if (!name) return [];
        out.push({
          check: name,
          column: data[`column.${at}`] ?? "",
          message: data[`message.${at}`] ?? "",
        });
      }
      return out;
    }
  } catch {
    return [];
  }
  return [];
}
